### PR TITLE
Add multi lathe

### PR DESCRIPTION
**Pull request by Volence**
Precision Industrial Lathe Design

The Precision Industrial Lathe is a new multiblock meant to replace the Lathe functionality of the Large Processing Factory, as well as give a slight increase of usefulness of lathe recipes for a higher cost (only unlocked with osmium) (this latter functionality can also be removed if thought to be too powerful).

It can be built in 2 ways in order to give a bit of freedom of choice to set up due to it not being fully symmetrical:

With the grate casing to the right:

image

Or the grate casing to the left:

image

Inputs and outputs are only allowed on the 9 steel casings on each side, muffler/maintenance/energy hatches are allowed on any of the steel casings:

image

With this multiblock, a new type of tiering item is introduced, the Item Pipe Casings, which are similar to the Pipe Casings, but based around item pipes as opposed to the fluid pipes. The idea is the higher tier you get the better it can handle the items being processed through it. For the machine there’s platinum, osmium, quantium, fluxed electrum, and black plutonium (but there’s also tin, brass, and electrum lower tiers for possible other machines in the future):

image

Osmium casings and above will trigger Precision Lathe Mode, which now allows 1 ingot/gem -> 2 rods instead of 1 rod + 2 small dust:

image
image

For parallels and speed bonuses, we have the following table:

        Material: Parallel Bonus, Speed Bonus
        Platinum: 4,  1
        Osmium: 8, 1.25
        Quantium: 12, 1.5
        FluxedElectrum: 16, 2
        BlackPlutonium: 32, 2.5

The calculation for the speed bonuses are:
1 / (Speed Boost + Voltage Tier / 4F)

The calculation for Parallels is the following:
Pipe Parallel Bonus + (Voltage Tier * 2)

### CODE REVIEW

**Comment by boubou19** [comment]
@Volence i see that you converted lots of imports into star imports, can you revert this to stick with explicit imports pls? Also you added tons of new recipes to populate the recipe map of the multiblock lathe, but you can just take them from the single block recipe map directly. See how gt++ electrolyzer or centrifuge do it for instance. Adding recipes to the multiblock recipe map should only be done if someone wants to add multiblock exclusive recipes.

**Comment by Volence** [comment]
> @Volence i see that you converted lots of imports into star imports, can you revert this to stick with explicit imports pls? Also you added tons of new recipes to populate the recipe map of the multiblock lathe, but you can just take them from the single block recipe map directly. See how gt++ electrolyzer or centrifuge do it for instance. Adding recipes to the multiblock recipe map should only be done if someone wants to add multiblock exclusive recipes.

Oh sorry about that, just found out it's my editor, I'll fix those imports. As for the recipes, they're made to be an alternative once you get the Osmium component (and as stated, could change if this is too powerful). It now goes 1 ingot/gem to 2 rods instead of 1 rod and 2 small dusts. The idea is just an upgrade/alternative since the extruder basically makes using ingots in a lathe useless after MV, and a bit of a buff for using gems at that recipe.

image

**Comment by LazyFleshWasTaken** [comment]
Will the precision lathe eliminate all dust byproduct, or just for rods? Because there are also recipes such as chipped/flawed gems to bolts, and flawless to long rods, exquisite to lenses, plates to lenses, and wood lathing that all produce the item and a byproduct of dust.

**Comment by Volence** [comment]
> Will the precision lathe eliminate all dust byproduct, or just for rods? Because there are also recipes such as chipped/flawed gems to bolts, and flawless to long rods, exquisite to lenses, plates to lenses, and wood lathing that all produce the item and a byproduct of dust.

Only thing changed was Rods/Gems -> 2 Rods, everything else in it stays the same as the single block recipe

**Comment by boubou19** [comment]
> As for the recipes, they're made to be an alternative once you get the Osmium component (and as stated, could change if this is too powerful). It now goes 1 ingot/gem to 2 rods instead of 1 rod and 2 small dusts. The idea is just an upgrade/alternative since the extruder basically makes using ingots in a lathe useless after MV, and a bit of a buff for using gems at that recipe.

I get this, but a lathe is very different from an extruder: while an extruder pushes the material into a shaped hole, which deforms the material without losing anything, the lathe physically removes material to the desired shape. This is why it makes no sense to remove the 2 small dusts, because they have a meaning. Also the lathe is used for all the gem materials, because you can't deform those to a desired shape.

**Comment by Volence** [comment]
> > As for the recipes, they're made to be an alternative once you get the Osmium component (and as stated, could change if this is too powerful). It now goes 1 ingot/gem to 2 rods instead of 1 rod and 2 small dusts. The idea is just an upgrade/alternative since the extruder basically makes using ingots in a lathe useless after MV, and a bit of a buff for using gems at that recipe.
> 
> I get this, but a lathe is very different from an extruder: while an extruder pushes the material into a shaped hole, which deforms the material without losing anything, the lathe physically removes material to the desired shape. This is why it makes no sense to remove the 2 small dusts, because they have a meaning. Also the lathe is used for all the gem materials, because you can't deform those to a desired shape.

Yeah that makes sense, I wasn't thinking hard enough with the realism in mind, thinking maybe once you get to a certain progression the technology could possibly be designed to handle these original pitfalls of lathe designs as we know them.

That being said, if we'd like to stick to the more realistic route to this and get rid of this addition to the machine, would we want to make change the ration to something like 2 gems/metals to 3 rods + 2 small dusts? If not I can gut any of the custom recipes and just have it be like the other normal multis and just be a stronger version of the single block

**Comment by chochem** [comment]
hmm. do we really want to add 8 casing blocks that can only be used in this one special multi?
We already have pipe casings and I dont anything anywhere says they are fluid specific.

**Comment by FourIsTheNumber** [comment]
> hmm. do we really want to add 8 casing blocks that can only be used in this one special multi? We already have pipe casings and I dont anything anywhere says they are fluid specific.

These are pretty generic, for what it's worth. Could easily reuse

**Comment by chochem** [comment]
I would also say that the recipe gen should be automatic like every other multi and not manual. Automatic does not mean that nothing can be adjusted. Though it sounds like boubou doesnt want you to adjust things anyway.

**Comment by Dream-Master** [comment]
> hmm. do we really want to add 8 casing blocks that can only be used in this one special multi? We already have pipe casings and I dont anything anywhere says they are fluid specific.

We could use it in other multies as well. If they fit better than existing ones I would choose the new casings.

**Comment by boubou19** [comment]
> > > As for the recipes, they're made to be an alternative once you get the Osmium component (and as stated, could change if this is too powerful). It now goes 1 ingot/gem to 2 rods instead of 1 rod and 2 small dusts. The idea is just an upgrade/alternative since the extruder basically makes using ingots in a lathe useless after MV, and a bit of a buff for using gems at that recipe.
> > 
> > 
> > I get this, but a lathe is very different from an extruder: while an extruder pushes the material into a shaped hole, which deforms the material without losing anything, the lathe physically removes material to the desired shape. This is why it makes no sense to remove the 2 small dusts, because they have a meaning. Also the lathe is used for all the gem materials, because you can't deform those to a desired shape.
> 
> Yeah that makes sense, I wasn't thinking hard enough with the realism in mind, thinking maybe once you get to a certain progression the technology could possibly be designed to handle these original pitfalls of lathe designs as we know them.
> 
> That being said, if we'd like to stick to the more realistic route to this and get rid of this addition to the machine, would we want to make change the ration to something like 2 gems/metals to 3 rods + 2 small dusts? If not I can gut any of the custom recipes and just have it be like the other normal multis and just be a stronger version of the single block

We could give it 2 ingots / gems to 2 rods and a full dust, it's just a built in x2 on the parallel... But even if we do this, i'd like to have a special conversion from single block recipe map to the multi, like the electrolyzer and the centrifuge are doing. But idk if it's really worth the hassle, as the lathe is used in LV for efficiency, then it's used later only to make the lenses and rods from gem. No need to make it competitive, as the usage of the machine is already niche in the pack, and people won't bother migrate their hundreds of rod patterns in a machine that yields back dusts from processing.

Really for me the only reason to have ths

**Comment by Volence** [comment]
> We could give it 2 ingots / gems to 2 rods and a full dust, it's just a built in x2 on the parallel... But even if we do this, i'd like to have a special conversion from single block recipe map to the multi, like the electrolyzer and the centrifuge are doing. But idk if it's really worth the hassle, as the lathe is used in LV for efficiency, then it's used later only to make the lenses and rods from gem. No need to make it competitive, as the usage of the machine is already niche in the pack, and people won't bother migrate their hundreds of rod patterns in a machine that yields back dusts from processing.
> 
> Really for me the only reason to have this multi is to reduce the need for a PA or a LPF, so we have a dedicated multi, with some parallelism and a base recipe boost.

Yeah, the implementation for the recipe changes was done poorly, chochem helped point me in a better direction in the discord. That being said, before I go fixing it, just wondering if it's just something I should revert before spending extra time on it. The original intent was just to check off the LPF items, but the recipe change was added as a slight change to spice things up from the typical "here's just a multi that goes faster", although it's also fine just being just a dedicated multi from the LPF.

"as the usage of the machine is already niche in the pack" - this was also thought of when coming up with it, as I was hoping to make it so it's worth having an option/alternative to the standards that everyone goes for in their lab. There's the benefit of not having to deal with all the small dusts too.

Just a small question, for "and people won't bother migrate their hundreds of rod patterns in a machine that yields back dusts from processing." do you mean current players who are far in the pack and have it currently, or players when they get to that point? I'm not sure people will be having hundreds of patterns when they unlock this part of the machine

**Comment by boubou19** [comment]
> Yeah, the implementation for the recipe changes was done poorly, chochem helped point me in a better direction in the discord. That being said, before I go fixing it, just wondering if it's just something I should revert before spending extra time on it. The original intent was just to check off the LPF items, but the recipe change was added as a slight change to spice things up from the typical "here's just a multi that goes faster", although it's also fine just being just a dedicated multi from the LPF.

Imo just revert the new recipes and make it so the multiblock recipe map is populated from the single block one. We can always go back on this later, if really the community is pushing for it.

> Just a small question, for "and people won't bother migrate their hundreds of rod patterns in a machine that yields back dusts from processing." do you mean current players who are far in the pack and have it currently, or players when they get to that point? I'm not sure people will be having hundreds of patterns when they unlock this part of the machine

By this i mean that currently established players will never go from the multiblock multi to the lathe multi for the rods. And for those before this stage of the progression, i suppose the single block extruder is already automated for metallic rods.

**Comment by boubou19** [comment]
they are fine as is, and we can always tune them later anyways.

**Comment by Volence** [comment]
Alright @boubou19, just reverted the recipe map changes, fixed the star imports, and changed the machine to use the regular lathe recipe map.

File: src/main/java/gregtech/common/blocks/GT_Block_Casings11.java [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+        ItemList.Casing_Item_Pipe_Platinum.set(new ItemStack(this, 1, 3));
+        ItemList.Casing_Item_Pipe_Osmium.set(new ItemStack(this, 1, 4));
+        ItemList.Casing_Item_Pipe_Quantium.set(new ItemStack(this, 1, 5));
+        ItemList.Casing_Item_Pipe_Fluxed_Electrum.set(new ItemStack(this, 1, 6));
+        ItemList.Casing_Item_Pipe_Black_Plutonium.set(new ItemStack(this, 1, 7));
+    }
+
+    @Override
+    public int getTextureIndex(int aMeta) {
+        return (16 << 7) | (aMeta + 48);
```
Comment (FourIsTheNumber) [inline src/main/java/gregtech/common/blocks/GT_Block_Casings11.java:41]:
```suggestion
        return (16 << 7) | (aMeta + 64);
```
Overlapping with BC10, breaks all of those casings

Comment (NotAPenguin0) [reply src/main/java/gregtech/common/blocks/GT_Block_Casings11.java:41]: Please also document allocating the texture page in Textures.java :)

**Review by boubou19 — APPROVED** [review]

**Pull request merged**